### PR TITLE
CRUD API for a todolist

### PR DESCRIPTION
**Pull request by BhujelVictor**
# Pull Request

## Description
API that performs CRUD operation

## Changes Made
- Added CRUD API files for a todo app.

## Screenshots 
TaskListView

TaskListView

TaskAddView

TaskAddView

TaskDeleteView

TaskDeleteView

TaskListViewAfter-Deletion-task3

TaskListViewAfter-Deletion-task3

TaskUpdateView

TaskUpdateView

TaskListViewAfterUpdate-on-task3

TaskListViewAfterUpdate-on-task3

### CODE REVIEW

File: ToDo/serializers.py [hunk 1]
```diff
@@ -0,0 +1,9 @@
+from rest_framework import serializers
+from .models import ToDoList
+
+class TaskSerializer(serializers.ModelSerializer):
+    class Meta:
+        model = ToDoList
+        fields = ['id', 'title', 'due_date', 'created_date', 'completed']
```
Comment (Biplav-00) [inline ToDo/serializers.py]: rename it to is_complete

File: ToDo/urls.py [hunk 2]
```diff
@@ -0,0 +1,12 @@
+from django.urls import path
+from ToDo.views import HomeAPIView, TaskListView, TaskDetailView, TaskAddView, TaskUpdateView, TaskDeleteView
+
+urlpatterns = [
+    path('', HomeAPIView.as_view(), name='home'),
```
Comment (Biplav-00) [inline ToDo/urls.py]: there are generally two view class
Task list view and task detail view

task list view consists of  (list task and post task)
task detail view consists of (get individual task by id , update and delete)

File: ToDo/views.py [hunk 3]
```diff
@@ -0,0 +1,78 @@
+from ToDo.models import ToDoList
```
Comment (Biplav-00) [inline ToDo/views.py]: rename it to Todo

File: ToDo/views.py [hunk 4]
```diff
@@ -0,0 +1,78 @@
...
+        urls = {
+            'List':'list/',
+            'Detail':'detail/<int:id>',
+            'Add':'add-task/',
+            'Update':'update/task',
+            'Delete':'delete-task/<int:id>',
+        }
+        return Response(urls)
+
+class TaskListView(APIView):
```
Comment (Biplav-00) [inline ToDo/views.py]: do changes by above comment

File: accounts/urls.py [hunk 5]
```diff
@@ -0,0 +1,9 @@
+from django.urls import path
+from accounts.views import *
+urlpatterns = [
+    path('api/register/', UserRegistrationView.as_view(), name='register-user'),
```
Comment (Biplav-00) [inline accounts/urls.py]: change register to register_user

File: accounts/urls.py [hunk 6]
```diff
@@ -0,0 +1,9 @@
...
+    path('api/login/', UserLoginView.as_view(), name='login-user'),
+    path('api/verify-email/', VerifyEmailView.as_view(), name='verify-email'),
```
Comment (Biplav-00) [inline accounts/urls.py]: change verify-email to verify-email

File: accounts/urls.py [hunk 7]
```diff
@@ -0,0 +1,9 @@
...
+    path('api/resend-code/', ResendVerificationCodeView.as_view(), name='resend-code')
```
Comment (Biplav-00) [inline accounts/urls.py]: same here

File: accounts/views.py [hunk 8]
```diff
@@ -0,0 +1,84 @@
+from rest_framework.response import Response
+from rest_framework import status
+from rest_framework.views import APIView
+from accounts.serializers import *
+from django.contrib.auth import authenticate
+from django.core.mail import send_mail
+import random
+
+def generate_otp():
```
Comment (Biplav-00) [inline accounts/views.py]: can we add this generate_otp() to service class and call it in the view

File: accounts/views.py [hunk 9]
```diff
@@ -0,0 +1,84 @@
...
+    otp_code = str(random.randint(100000, 999999))
+    return otp_code
+
+class UserRegistrationView(APIView):
+    def post(self, request, format=None):
+        serializer = UserRegistrationSerializer(data=request.data)
```
Comment (Biplav-00) [inline accounts/views.py:15]: i think it is better approach to shift the db related stuff to service class

File: restToDo/asgi.py [hunk 10]
```diff
@@ -0,0 +1,16 @@
+"""
```
Comment (Biplav-00) [inline restToDo/asgi.py]: rename restTodo to rest_todo

File: restToDo/urls.py [hunk 11]
```diff
@@ -0,0 +1,24 @@
+"""
+URL configuration for restToDo project.
```
Comment (Biplav-00) [inline restToDo/urls.py:2]: can we remove unnecessary comments from all the code

only include the comments that are necessary

File: restToDo/urls.py [hunk 12]
```diff
@@ -0,0 +1,24 @@
...
+
+The `urlpatterns` list routes URLs to views. For more information please see:
+    https://docs.djangoproject.com/en/5.0/topics/http/urls/
+Examples:
```
Comment (Biplav-00) [inline restToDo/urls.py:6]: same here

File: restToDo/wsgi.py [hunk 13]
```diff
@@ -0,0 +1,16 @@
+"""
+WSGI config for restToDo project.
+
+It exposes the WSGI callable as a module-level variable named ``application``.
```
Comment (Biplav-00) [inline restToDo/wsgi.py]: same in all the files

File: ToDo/apps.py [hunk 14]
```diff
@@ -0,0 +1,6 @@
+from django.apps import AppConfig
+
+
+class TodoConfig(AppConfig):
+    default_auto_field = 'django.db.models.BigAutoField'
+    name = 'ToDo'
```
Comment (Biplav-00) [inline ToDo/apps.py]: rename it to todo

File: ToDo/models.py [hunk 15]
```diff
@@ -0,0 +1,21 @@
...
+
+
+# to add a default due time while creating a todo item
+def one_week_hence():
+    return timezone.now() + timezone.timedelta(days=7)
+
+#defining attribute of the list
+class ToDoList(models.Model):
+    title = models.CharField(max_length=100, unique=True)
+    created_date = models.DateTimeField(auto_now_add=True)
```
Comment (Biplav-00) [inline ToDo/models.py]: rename it to created_at

File: ToDo/models.py [hunk 16]
```diff
@@ -0,0 +1,21 @@
...
+    due_date = models.DateTimeField(default=one_week_hence)
```
Comment (Biplav-00) [inline ToDo/models.py]: rename it to deadline or something meaningful name

**Review by Biplav-00 — CHANGES_REQUESTED** [review]

File: accounts/views.py [hunk 17]
```diff
@@ -0,0 +1,87 @@
...
+from rest_framework import status
+from rest_framework.views import APIView
+from accounts.serializers import *
+from django.contrib.auth import authenticate, login, logout
+from django.core.mail import send_mail
+from accounts.services import GenOTP # for OTP generation
+from accounts.authentications import CustomEmailBackend
+from rest_framework.permissions import AllowAny
+from rest_framework.permissions import IsAuthenticated
+class UserRegistrationView(APIView):
```
Comment (yonjans) [inline accounts/views.py]: Add 2 line breaks.

File: accounts/serializers.py [hunk 18]
```diff
@@ -0,0 +1,70 @@
...
+    
+    def validate(self, data):
+        email = data.get('email')
+        otp = data.get('otp')
+        #Checking if email exists or not
+        if User.objects.filter(email=email).exists():
+            user = User.objects.get(email=email)
+            #Checking for otp record
+            try:
+                otp_record = OTP.objects.get(user=user, otp=otp, otp_created_at__gte=timezone.now() - timedelta(minutes=5))
```
Comment (yonjans) [inline accounts/serializers.py]: What is the logic here? I think it should be fixed.

File: accounts/serializers.py [hunk 19]
```diff
@@ -0,0 +1,70 @@
...
+            except OTP.DoesNotExist:
+                raise serializers.ValidationError('OTP does not match or has timed out')
+
+            otp_record.delete()
```
Comment (yonjans) [inline accounts/serializers.py]: Do we need to remove record?

**Review by yonjans — CHANGES_REQUESTED** [review]

**Pull request merged**